service/collector/elb.go

```diff
@@ -139,31 +146,70 @@ func (e *ELB) collectForAccount(ch chan<- prometheus.Metric, awsClients clientaw
 
 	var lbs []loadBalancer
 	{
-		i := &elb.DescribeTagsInput{}
-		for _, l := range loadBalancers {
-			i.LoadBalancerNames = append(i.LoadBalancerNames, l.LoadBalancerName)
+		// AWS API has a limit for maximum number of LoadBalancerNames in
+		// single Describe request so it must be done in batches of
+		// maxELBsInOneBatch. In order to not spend so much time on this,
+		// perform requests concurrently and synchronize them with errgroup.
+		errGroup, _ := errgroup.WithContext(context.TODO())
+		// Slice for ELB tag description results.
+		var tagOutputs []*elb.DescribeTagsOutput
+
+		lbNames := loadBalancerNames
+		mutex := &sync.Mutex{}
+		for len(lbNames) > 0 {
+			batchSize := maxELBsInOneDescribeTagsBatch
+			if len(lbNames) < batchSize {
+				batchSize = len(lbNames)
+			}
+
+			tagInput := &elb.DescribeTagsInput{
+				LoadBalancerNames: lbNames[0:batchSize],
+			}
+			lbNames = lbNames[batchSize:]
+
+			errGroup.Go(func() error {
+				o, err := awsClients.ELB.DescribeTags(tagInput)
+				if err != nil {
+					return microerror.Mask(err)
+				}
+
+				mutex.Lock()
+				tagOutputs = append(tagOutputs, o)
+				mutex.Unlock()
+
+				return nil
+			})
 		}
 
-		o, err := awsClients.ELB.DescribeTags(i)
+		// Now wait for all requests to complete.
+		err := errGroup.Wait()
 		if err != nil {
 			return microerror.Mask(err)
 		}
 
-		for _, d := range o.TagDescriptions {
-			lb := loadBalancer{
-				Name: *d.LoadBalancerName,
-				Tags: make(map[string]string),
+		// Extract tags from responses and create further loadBalancer types
+		// based on these.
+		for _, o := range tagOutputs {
+			if o == nil {
```

Comment:
I think when we have sucha condition this means a bug in the code above. I'd prefer to have a panic and fix it.

Reply:
Oh, true. With earlier implementation this was necessary due to different cases whether `numLBs % batchSize == 0` was true or not. I'll remove the check so it will panic in case it behaves wrong.